derive.h

```diff
@@ -0,0 +1,311 @@
+#include <iostream>
+#include <cassert>
+#include <memory>
+#include <optional>
+
+using std::ostream;
+using std::shared_ptr;
+using std::optional;
+
+namespace detail {
+enum Operation {
+	plus, 
+	minus,
+	mult,
+	power
+};
+
+
+ostream& operator<<(ostream& os, const Operation& operation) {
+	switch (operation) {
+		case plus: os << "+"; break;
+		case minus: os << "-"; break;
+		case mult: os << "*"; break;
+		case power: os << "^"; break;
+		default: assert(0); 
+	}
+	return os;
+}
+
+class Expression {
```

Comment:
Base class with virtual functions should always have a virtual destructor
```
virtual ~Expression() {}
```

Reply:
You are right. Added.